pkg/blobcache/blobcache.go

```diff
@@ -0,0 +1,396 @@
+			filename := filepath.Join(directory, makeFilename(*instanceDigest, false))
+			manifestBytes, err := ioutil.ReadFile(filename)
+			if err == nil {
+				return manifestBytes, manifest.GuessMIMEType(manifestBytes), nil
+			}
+			if !os.IsNotExist(err) {
+				return nil, "", errors.Wrapf(err, "error checking for manifest file %q", filename)
+			}
+		}
+	}
+	return s.source.GetManifest(ctx, instanceDigest)
+}
+
+func (s *blobCacheSource) GetBlob(ctx context.Context, blobinfo types.BlobInfo) (io.ReadCloser, int64, error) {
+	present, size, err := s.reference.HasBlob(blobinfo)
+	if err != nil {
+		return nil, -1, err
+	}
+	if present {
+		for _, directory := range s.reference.directories {
+			for _, isConfig := range []bool{false, true} {
+				filename := filepath.Join(directory, makeFilename(blobinfo.Digest, isConfig))
+				f, err := os.Open(filename)
+				if err == nil {
+					return f, size, nil
+				}
+				if !os.IsNotExist(err) {
+					return nil, -1, errors.Wrapf(err, "error checking for cache file %q", filepath.Join(directory, filename))
+				}
+			}
+		}
+	}
+	rc, size, err := s.source.GetBlob(ctx, blobinfo)
+	if err != nil {
+		return rc, size, errors.Wrapf(err, "error reading blob from source image %q", transports.ImageName(s.reference))
+	}
+	return rc, size, nil
+}
+
+func (s *blobCacheSource) GetSignatures(ctx context.Context, instanceDigest *digest.Digest) ([][]byte, error) {
+	return s.source.GetSignatures(ctx, instanceDigest)
+}
+
+func (s *blobCacheSource) LayerInfosForCopy(ctx context.Context) ([]types.BlobInfo, error) {
+	return nil, nil
+}
+
+func (d *blobCacheDestination) Reference() types.ImageReference {
+	return d.reference
+}
+
+func (d *blobCacheDestination) Close() error {
+	return d.destination.Close()
+}
+
+func (d *blobCacheDestination) SupportedManifestMIMETypes() []string {
+	return d.destination.SupportedManifestMIMETypes()
+}
+
+func (d *blobCacheDestination) SupportsSignatures(ctx context.Context) error {
+	return d.destination.SupportsSignatures(ctx)
+}
+
+func (d *blobCacheDestination) DesiredLayerCompression() types.LayerCompression {
+	return d.destination.DesiredLayerCompression()
+}
+
+func (d *blobCacheDestination) AcceptsForeignLayerURLs() bool {
+	return d.destination.AcceptsForeignLayerURLs()
+}
+
+func (d *blobCacheDestination) MustMatchRuntimeOS() bool {
+	return d.destination.MustMatchRuntimeOS()
+}
+
+func (d *blobCacheDestination) IgnoresEmbeddedDockerReference() bool {
+	return d.destination.IgnoresEmbeddedDockerReference()
+}
+
+func (d *blobCacheDestination) PutBlob(ctx context.Context, stream io.Reader, inputInfo types.BlobInfo, isConfig bool) (types.BlobInfo, error) {
+	var tempfile *os.File
+	var err error
+	if inputInfo.Digest != "" {
+		directory := d.reference.directories[0]
+		if directory == "" {
+			directory = "."
+		}
+		filename := filepath.Join(directory, makeFilename(inputInfo.Digest, isConfig))
+		tempfile, err = ioutil.TempFile(directory, makeFilename(inputInfo.Digest, isConfig))
```

Comment:
(Previously `directories[0] == ""` meant “create a temporary directory”; that has intentionally gone away?)

- Is it a good default to store temporary caches to `"."`? It’s not _impossible_ but it’s a bit surprising.
- Does `directories == []string{""}` work consistently? AFAICS `StringWithinTransport` will format it as (`blob-cache:`)`:`_ref_, and `ParseReference` interprets that as `[]string{}`, not `[]string{""}`.

Reply:
Now that we can only create caches using `NewBlobCache()`, the directory (of which there can only be one) must be specified, or we return an error.